File: debs/CPU/Settings/FRootListController.m
#import <Preferences/PSListController.h>
#import <Preferences/PSSpecifier.h>
#import <UIKit/UIKit.h>
#import <spawn.h>
#import <sys/wait.h>
#import <notify.h>
#import <dlfcn.h>
#import <CPUthermalPaths.h>
#import <CPUthermalFrequency.h>

// ============================================================
// 注意: 禁止使用 @"" ObjC 字符串常量
// roothide 重映射会破坏 __cfstring 内部指针，导致 SIGBUS
// 所有字符串通过 C 字符串 + stringWithUTF8String: 动态创建
// ============================================================

@interface FRootListController : PSListController
@end

@implementation FRootListController

- (NSString *)prefPath {
    return CPUthermalCurrentPrefPath();
}

- (NSMutableDictionary *)prefs {
    NSMutableDictionary *d = CPUthermalReadMutablePrefs();
    if (!d) d = [NSMutableDictionary dictionary];
    return d;
}

- (void)savePrefs:(NSMutableDictionary *)prefs {
    CPUthermalWritePrefs(prefs);
    notify_post(kCPUthermalSettingsChangedNotifC);
}

- (void)setPreferenceValue:(id)value specifier:(PSSpecifier *)spec {
    NSString *key = [spec propertyForKey:S("key")];
    if (!key) return;

    NSMutableDictionary *prefs = [self prefs];

    // 功率模式统一规范为字符串，防止框架传入 NSNumber 索引
    if ([key isEqualToString:S("powerMode")]) {
        if ([value isKindOfClass:[NSNumber class]]) {
            value = ([value intValue] == 0) ? S("lowPower") : S("fullPower");
        } else if (![value isKindOfClass:[NSString class]]) {
            value = S("fullPower");
        }
    }

    if ([key isEqualToString:S("lowPowerLockMHz")]) {
        value = [NSString stringWithFormat:S("%d"), CPUthermalLowPowerLockMHzFromValue(value)];
    }

    prefs[key] = value;
    CPUthermalWritePrefs(prefs);

    // 功率模式只发专用通知，避免 settingsChanged + powerModeChanged 重复应用。
    if ([key isEqualToString:S("powerMode")]) {
        notify_post(kCPUthermalPowerModeChangedNotifC);
    } else {
        notify_post(kCPUthermalSettingsChangedNotifC);
    }
}

- (id)readPreferenceValue:(PSSpecifier *)spec {
    NSString *key = [spec propertyForKey:S("key")];
    if (!key) return nil;
    
    // 功率模式为字符串值，默认满血(解除温控)
    if ([key isEqualToString:S("powerMode")]) {
        id mode = [self prefs][key];
        if (mode && [mode isKindOfClass:[NSString class]]) return mode;
        return S("fullPower");
    }


    if ([key isEqualToString:S("lowPowerLockMHz")]) {
        return [NSString stringWithFormat:S("%d"),
                CPUthermalLowPowerLockMHzFromValue([self prefs][key])];
    }
    
    // 解除温控高频目标 — 编辑框需返回字符串，未设置时返回空串
    if ([key isEqualToString:S("cpuMinPowerValue")]) {
        id freqVal = [self prefs][key];
        return freqVal ? [freqVal description] : S("");
    }

    id val = [self prefs][key];
    if (val) return val;

    // 特定防护功能默认开启（开箱即用免配置）
    if ([key isEqualToString:S("thermalBlockNotifPopup")] ||
        [key isEqualToString:S("thermalPreventDimmingEnabled")]) {
        return [NSNumber numberWithBool:YES];
    }
    return [NSNumber numberWithBool:NO]; // 全部默认关闭
}

#pragma mark - 工具方法

- (void)openURLString:(NSString *)urlString fallback:(NSString *)fallbackURL failureMessage:(NSString *)failureMessage {
    NSURL *url = [NSURL URLWithString:urlString];
    if (!url) return;

    [[UIApplication sharedApplication] openURL:url
                                       options:[NSDictionary dictionary]
                             completionHandler:^(BOOL success) {
        if (success) return;
        if (fallbackURL) {
            NSURL *fallback = [NSURL URLWithString:fallbackURL];
            if (fallback) {
                [[UIApplication sharedApplication] openURL:fallback options:[NSDictionary dictionary] completionHandler:nil];
                return;
            }
        }
        if (failureMessage) {
            [self showSimpleAlertWithTitle:S("提示") message:failureMessage];
        }
    }];
}

- (void)showSimpleAlertWithTitle:(NSString *)title message:(NSString *)message {
    UIAlertController *alert = [UIAlertController
        alertControllerWithTitle:title
        message:message
        preferredStyle:UIAlertControllerStyleAlert];
    [alert addAction:[UIAlertAction actionWithTitle:S("好的") style:UIAlertActionStyleDefault handler:nil]];
    [self presentViewController:alert animated:YES completion:nil];
}

#pragma mark - 解除温控高频目标

/// 读取输入框当前文本：优先从 PSEditTextCell 实时读取（可能尚未写入 prefs），兜底从 prefs 读取
- (NSString *)currentCPUFrequencyInput {
    for (PSSpecifier *spec in self.specifiers) {
        NSString *key = [spec propertyForKey:S("key")];
        if (!key || ![key isEqualToString:S("cpuMinPowerValue")]) {
            continue;
        }
        if ([self respondsToSelector:@selector(cachedCellForSpecifier:)]) {
            UITableViewCell *cell = (UITableViewCell *)[self performSelector:@selector(cachedCellForSpecifier:) withObject:spec];
            if (cell) {
                @try {
                    UITextField *textField = (UITextField *)[cell valueForKey:S("textField")];
                    if ([textField isKindOfClass:[UITextField class]] && textField.text.length > 0) {
                        return textField.text;
                    }
                } @catch (NSException *exception) {
                    // KVC 失败则回退到 prefs
                }
            }
        }
        break;
    }

    id value = [self prefs][S("cpuMinPowerValue")];
    if (value && [value isKindOfClass:[NSString class]] && [(NSString *)value length] > 0) {
        return (NSString *)value;
    }
    return S("");
}

/// 「锁定频率」按钮动作：确认数值后立即应用
- (void)applyCPUFrequencyLock {
    NSString *input = [self currentCPUFrequencyInput];
    int mhz = [input intValue];
    if (mhz < kCPUthermalMinimumCPUFrequencyMHz || mhz > kCPUthermalMaximumCPUFrequencyMHz) {
        [self showSimpleAlertWithTitle:S("提示")
                              message:S("请输入 300 到 6000 之间的 CPU 目标频率（MHz）。")];
        return;
    }

    NSString *message = [NSString stringWithFormat:S("确认将解除温控高频目标设为 %d MHz？\n超过硬件上限会自动钳制到设备最大值。"), mhz];
    UIAlertController *alert = [UIAlertController
        alertControllerWithTitle:S("应用高频锁定")
        message:message
        preferredStyle:UIAlertControllerStyleAlert];
    [alert addAction:[UIAlertAction actionWithTitle:S("取消") style:UIAlertActionStyleCancel handler:nil]];
    [alert addAction:[UIAlertAction actionWithTitle:S("确定") style:UIAlertActionStyleDefault handler:^(UIAlertAction *action) {
        [self commitCPUFrequencyLock:mhz];
    }]];
    [self presentViewController:alert animated:YES completion:nil];
}

/// 提交锁定：写入 prefs → 发通知，由正在运行的插件立即重应用。
- (void)commitCPUFrequencyLock:(int)mhz {
    NSMutableDictionary *prefs = [self prefs];
    prefs[S("cpuMinPowerValue")] = [NSString stringWithFormat:S("%d"), mhz];
    [self savePrefs:prefs]; // savePrefs 内部已发 settingsChanged 通知

    [self showSimpleAlertWithTitle:S("已应用")
                          message:[NSString stringWithFormat:S("解除温控高频目标已设为 %d MHz；系统会自动选择最接近的可用 P-State。"), mhz]];
}

#pragma mark - 重启用户空间

- (void)usreboot {
    UIAlertController *alert = [UIAlertController
        alertControllerWithTitle:S("重启用户空间")
        message:S("重启 SpringBoard 和所有用户态进程？")
        preferredStyle:UIAlertControllerStyleAlert];
    [alert addAction:[UIAlertAction actionWithTitle:S("取消")
                                              style:UIAlertActionStyleCancel handler:nil]];
    [alert addAction:[UIAlertAction actionWithTitle:S("确定重启")
                                              style:UIAlertActionStyleDestructive handler:^(UIAlertAction *action) {
        pid_t pid = 0;
        NSString *toolPath = CPUthermalToolPath();
        if (toolPath.length > 0 && [[NSFileManager defaultManager] isExecutableFileAtPath:toolPath]) {
            char *args[] = {"CPUthermalTool", "userspace-reboot", NULL};
            if (posix_spawn(&pid, [toolPath fileSystemRepresentation], NULL, NULL, args, NULL) == 0) {
                waitpid(pid, NULL, 0);
                return;
            }
        }

        NSString *launchctlPath = CPUthermalLaunchctlPath();
        if (launchctlPath.length > 0) {
            char *args[] = {"launchctl", "reboot", "userspace", NULL};
            posix_spawn(&pid, [launchctlPath fileSystemRepresentation], NULL, NULL, args, NULL);
            waitpid(pid, NULL, 0);
        }
    }]];
    [self presentViewController:alert animated:YES completion:nil];
}

#pragma mark - 开源代码

- (void)openSourceCode {
    [self openURLString:S("https://github.com/be-huge/insulation") fallback:nil failureMessage:S("无法打开 GitHub，请手动访问 https://github.com/be-huge/insulation")];
}

#pragma mark - Specifier 加载

- (NSArray *)specifiers {
    if (!_specifiers) {
        // 直接从 Root.plist 加载配置结构，Preferences 框架会自动正确解析 PSSegmentCell
        _specifiers = [self loadSpecifiersFromPlistName:S("Root") target:self];
    }
    return _specifiers;
}

@end

File: debs/CPU/Settings/FRootListController.swift
import Foundation
import Preferences
import UIKit

@objc(FRootListController)
public final class FRootListController: PSListController {
    private var prefPath: String { CPUthermalShared.currentPrefPath }
    private var legacyPrefPath: String? { CPUthermalShared.legacyPrefPaths.first }

    private func ensurePrefsDirectory() {
        CPUthermalShared.ensurePrefDirectory()
    }

    private func migrateLegacyPrefsIfNeeded() {
        _ = CPUthermalShared.readPrefs()
    }

    private func prefs() -> NSMutableDictionary {
        CPUthermalShared.readMutablePrefs()
    }

    private func savePrefs(_ prefs: NSMutableDictionary) {
        CPUthermalShared.writePrefs(prefs)
        notify_post(CPUthermalShared.settingsChangedNotification)
    }

    private func powerModeValue() -> String {
        prefs()["powerMode"] as? String ?? "fullPower"
    }

    private func powerModeTitle(_ mode: String) -> String {
        mode == "lowPower" ? "低功耗" : "防温控"
    }

    private func powerModeLabel() -> String {
        "功率模式：\(powerModeTitle(powerModeValue()))"
    }

    private func restartThermalmonitord() {
        CPUthermalShared.restartThermalmonitordSoon()
    }

    private func savePowerMode(_ mode: String?) {
        let prefs = self.prefs()
        prefs["powerMode"] = mode ?? "fullPower"
        savePrefs(prefs)
        notify_post(CPUthermalShared.powerModeChangedNotification)
        restartThermalmonitord()

        if let specifier = specifier(forID: "powerMode") {
            specifier.name = powerModeLabel()
        }
        reloadSpecifierID("powerMode", animated: true)
    }

    @objc public override func setPreferenceValue(_ value: Any?, specifier spec: PSSpecifier!) {
        guard let key = spec.property(forKey: "key") as? String else { return }
        let prefs = self.prefs()
        prefs[key] = value
        savePrefs(prefs)
    }

    @objc public override func readPreferenceValue(_ spec: PSSpecifier!) -> Any? {
        guard let key = spec.property(forKey: "key") as? String else { return nil }
        if let value = prefs()[key] {
            return value
        }
        if key == "keepCPMSAlive" {
            return true
        }
        if key == "suppressThermalNotifications" {
            return false
        }
        return true
    }

    @objc public func readPowerModeValue(_ spec: PSSpecifier) -> Any {
        powerModeLabel()
    }

    @objc public func openPowerModePicker() {
        showPowerModePicker()
    }

    public override func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        let index = index(for: indexPath)
        if index >= 0, index < specifiers.count,
           let specifier = specifiers[index] as? PSSpecifier,
           (specifier.property(forKey: "key") as? String) == "powerMode" {
            tableView.deselectRow(at: indexPath, animated: true)
            showPowerModePicker()
            return
        }
        super.tableView(tableView, didSelectRowAt: indexPath)
    }

    private func showPowerModePicker() {
        let alert = UIAlertController(
            title: "功率模式",
            message: "低功耗 = 限制 CPU 最高 2016MHz，不抬高待机频率\n防温控 = 减少降频/降亮度，保留系统安全保护",
            preferredStyle: .actionSheet
        )

        let currentMode = powerModeValue()
        let low = UIAlertAction(title: "低功耗", style: .default) { [weak self] _ in
            self?.savePowerMode("lowPower")
        }
        let full = UIAlertAction(title: "防温控", style: .default) { [weak self] _ in
            self?.savePowerMode("fullPower")
        }
        if currentMode == "lowPower" {
            low.setValue(true, forKey: "checked")
        } else {
            full.setValue(true, forKey: "checked")
        }

        alert.addAction(low)
        alert.addAction(full)
        alert.addAction(UIAlertAction(title: "取消", style: .cancel))

        if let popover = alert.popoverPresentationController {
            popover.sourceView = view
            popover.sourceRect = view.bounds
            popover.permittedArrowDirections = []
        }
        present(alert, animated: true)
    }

    private func openURLString(_ urlString: String, fallback: String?, failureMessage: String? = nil) {
        guard let url = URL(string: urlString) else { return }
        UIApplication.shared.open(url, options: [:]) { [weak self] success in
            guard !success else { return }
            if let fallback, let fallbackURL = URL(string: fallback) {
                UIApplication.shared.open(fallbackURL, options: [:])
                return
            }
            if let failureMessage {
                self?.showSimpleAlert(title: "提示", message: failureMessage)
            }
        }
    }

    private func showSimpleAlert(title: String, message: String) {
        let alert = UIAlertController(title: title, message: message, preferredStyle: .alert)
        alert.addAction(UIAlertAction(title: "好的", style: .default))
        present(alert, animated: true)
    }

    @objc public func openQQFeedbackGroup() {
        openURLString("https://qm.qq.com/q/JvllAQiEwI", fallback: nil)
    }

    @objc public func openAlipayDonate() {
        openURLString(
            "alipays://platformapi/startapp?appId=20000067&url=https%3A%2F%2Fqr.alipay.com%2Ffkx16683ylwdrfdo8fiuy01",
            fallback: "https://qr.alipay.com/fkx16683ylwdrfdo8fiuy01"
        )
    }

    @objc public func openRepo() {
        openURLString("sileo://source/https://huayuarc.github.io", fallback: "https://huayuarc.github.io")
    }

    @objc public func usreboot() {
        let alert = UIAlertController(
            title: "重启用户空间",
            message: "重启 SpringBoard 和所有用户态进程？",
            preferredStyle: .alert
        )
        alert.addAction(UIAlertAction(title: "取消", style: .cancel))
        alert.addAction(UIAlertAction(title: "确定重启", style: .destructive) { _ in
            if let toolPath = CPUthermalShared.toolPath,
               FileManager.default.isExecutableFile(atPath: toolPath),
               CPUthermalShared.runAndWait(path: toolPath, arguments: ["CPUthermalTool", "userspace-reboot"]) {
                return
            }
            _ = CPUthermalShared.runAndWait(path: CPUthermalShared.launchctlPath, arguments: ["launchctl", "reboot", "userspace"])
        })
        present(alert, animated: true)
    }

    private func switchSpecifier(label: String, key: String) -> PSSpecifier {
        let spec = PSSpecifier.preferenceSpecifierNamed(
            label,
            target: self,
            set: #selector(setPreferenceValue(_:specifier:)),
            get: #selector(readPreferenceValue(_:)),
            detail: nil,
            cell: .switchCell,
            edit: nil
        )
        spec?.identifier = key
        spec?.setProperty(key, forKey: "key")
        return spec!
    }

    private func powerModeSpecifier() -> PSSpecifier {
        let spec = PSSpecifier.preferenceSpecifierNamed(
            powerModeLabel(),
            target: self,
            set: nil,
            get: nil,
            detail: nil,
            cell: .buttonCell,
            edit: nil
        )
        spec?.identifier = "powerMode"
        spec?.setProperty("powerMode", forKey: "key")
        spec?.buttonAction = #selector(openPowerModePicker)
        return spec!
    }

    private func buttonSpecifier(label: String, action: Selector, identifier: String) -> PSSpecifier {
        let spec = PSSpecifier.preferenceSpecifierNamed(
            label,
            target: self,
            set: nil,
            get: nil,
            detail: nil,
            cell: .buttonCell,
            edit: nil
        )
        spec?.buttonAction = action
        spec?.identifier = identifier
        return spec!
    }

    public override var specifiers: NSMutableArray! {
        get {
            if let cached = value(forKey: "_specifiers") as? NSMutableArray {
                return cached
            }

            let specs = NSMutableArray()
            var group: PSSpecifier

            group = PSSpecifier.emptyGroup()!
            group.setProperty("CPUthermal", forKey: "label")
            specs.add(group)
            specs.add(switchSpecifier(label: "启用 CPUthermal", key: "enabled"))

            group = PSSpecifier.emptyGroup()!
            group.setProperty("功率模式", forKey: "label")
            group.setProperty("低功耗只限制最高频率，不再强制锁 2016MHz；防温控会减少降频/降亮度，但保留系统过热保护。", forKey: "footerText")
            specs.add(group)
            specs.add(powerModeSpecifier())

            group = PSSpecifier.emptyGroup()!
            group.setProperty("核心保护", forKey: "label")
            group.setProperty("建议保持默认：CPU/亮度保护开启，避免误判温度。", forKey: "footerText")
            specs.add(group)
            specs.add(switchSpecifier(label: "CPU 性能保护", key: "cpuProtection"))
            specs.add(switchSpecifier(label: "屏幕亮度保护", key: "brightnessProtection"))
            specs.add(switchSpecifier(label: "屏蔽高温通知", key: "suppressThermalNotifications"))

            group = PSSpecifier.emptyGroup()!
            group.setProperty("高级", forKey: "label")
            group.setProperty("强烈建议开启：温度超过 70°C 或读温失败时放行系统温控，防止异常发热和自动黑屏。", forKey: "footerText")
            specs.add(group)
            specs.add(switchSpecifier(label: "保留 CPMS 紧急保护", key: "keepCPMSAlive"))

            group = PSSpecifier.emptyGroup()!
            group.setProperty("操作", forKey: "label")
            specs.add(group)
            specs.add(buttonSpecifier(label: "重启用户空间", action: #selector(usreboot), identifier: "usreboot"))

            group = PSSpecifier.emptyGroup()!
            group.setProperty("关于我 / 投喂", forKey: "label")
            specs.add(group)
            specs.add(buttonSpecifier(label: "📮 QQ 交流反馈群", action: #selector(openQQFeedbackGroup), identifier: "qqGroup"))
            specs.add(buttonSpecifier(label: "💰 支付宝🧧打赏", action: #selector(openAlipayDonate), identifier: "alipayDonate"))
            specs.add(buttonSpecifier(label: "📦 Sileo 添加源", action: #selector(openRepo), identifier: "sileoRepo"))

            self.specifiers = specs
            return specs
        }
        set {
            super.specifiers = newValue
        }
    }
}
